File: ruby/minesweeper/minesweeper.rb
# frozen_string_literal: true

module Board
  ADJACENT_COORDINATES = lambda do |y_index, x_index|
    [x_index - 1, x_index + 1, y_index - 1, y_index + 1]
  end
  private_constant :ADJACENT_COORDINATES
  BLANK = " "
  private_constant :BLANK
  MINE = "*"
  private_constant :MINE
  HORIZONTAL_BORDER_START = "+"
  private_constant :HORIZONTAL_BORDER_START
  VERTICAL_BORDER = "|"
  private_constant :VERTICAL_BORDER

  module_function

  def transform(board)
    raise ArgumentError unless all_lines_same_length?(board)

    board
      .each
      .with_index
      .with_object(board)
      .map(&method(:transform_line))
  end

  def transform_line((line, y_index), board)
    return line if line.start_with?(HORIZONTAL_BORDER_START)
    raise ArgumentError unless line.start_with?(VERTICAL_BORDER)

    line
      .chars
      .each
      .with_index
      .with_object([y_index, board])
      .map(&method(:transform_character))
      .join
  end
  private_class_method :transform_line

  def transform_character((char, x_index), (y_index, board))
    return char if board_element?(char)
    raise ArgumentError unless char == BLANK

    adjacent_coordinates(board, y_index, x_index)
      .each
      .with_object(board)
      .sum(&method(:sum_mine))
      .then { |sum| sum.positive? ? sum : char }
  end

  def all_lines_same_length?(board)
    board.map(&:length).uniq.one?
  end
  private_class_method :all_lines_same_length?

  def board_element?(char)
    [VERTICAL_BORDER, MINE].include?(char)
  end
  private_class_method :board_element?

  def adjacent_coordinates(board, y_index, x_index)
    left, right, above, below = ADJACENT_COORDINATES.call(y_index, x_index)
    [[y_index, left], [y_index, right]].then do |coords|
      board[above] &&
        coords += [[above, left], [above, x_index], [above, right]]
      board[below] &&
        coords + [[below, left], [below, x_index], [below, right]]
    end
  end
  private_class_method :adjacent_coordinates

  def sum_mine(((y, x), board))
    board[y][x] == MINE ? 1 : 0
  end
  private_class_method :sum_mine
end

File: elixir/minesweeper/lib/minesweeper.ex
defmodule Minesweeper do
  @mine "*"
  @beginning_outer_bound -1

  @doc """
  Annotate empty spots next to mines with the number of mines next to them.
  """
  @spec annotate([String.t()]) :: [String.t()]

  def annotate([]), do: []

  def annotate(board) do
    board
    |> Enum.with_index()
    |> Enum.map(&annotate_line(board, &1))
  end

  defp annotate_line(board, {line, y_index}) do
    line
    |> String.codepoints()
    |> Enum.with_index()
    |> Enum.map(&annotate_character({y_index, board}, &1))
    |> Enum.join()
  end

  defp annotate_character({y_index, board}, {char, x_index}) do
    if char == @mine do
      char
    else
      board
      |> adjacent_coordinates(y_index, x_index)
      |> Enum.reduce(0, &sum_mine(board, &1, &2))
      |> substitute_character(char)
    end
  end

  defp adjacent_coordinates(board, y_index, x_index) do
    {left, right, above, below} = get_adjacent_coordinates(y_index, x_index)

    [{y_index, left}, {y_index, right}]
    |> concat_above_coordinates(board, {x_index, left, right, above})
    |> concat_below_coordinates(board, {x_index, left, right, below})
  end

  defp get_adjacent_coordinates(y_index, x_index) do
    {x_index - 1, x_index + 1, y_index - 1, y_index + 1}
  end

  defp concat_above_coordinates(
         coordinates,
         board,
         {x_index, left, right, above}
       ) do
    if above > @beginning_outer_bound and Enum.at(board, above) do
      [{above, x_index}, {above, right}]
      |> concat_left_coordinate(above, left)
      |> Enum.concat(coordinates)
    else
      coordinates
    end
  end

  defp concat_below_coordinates(
         coordinates,
         board,
         {x_index, left, right, below}
       ) do
    if Enum.at(board, below) do
      [{below, x_index}, {below, right}]
      |> concat_left_coordinate(below, left)
      |> Enum.concat(coordinates)
    else
      coordinates
    end
  end

  defp concat_left_coordinate(coordinates, y_direction, left) do
    if left > @beginning_outer_bound do
      [{y_direction, left} | coordinates]
    else
      coordinates
    end
  end

  defp sum_mine(board, {y, x}, acc) do
    char =
      board
      |> Enum.at(y)
      |> String.codepoints()
      |> Enum.at(x)

    if char == @mine, do: acc + 1, else: acc
  end

  defp substitute_character(sum, char), do: if(sum > 0, do: sum, else: char)
end
